docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,346 @@
+with the teardown in `lucet_context_backstop`. Both functions attempt to swap
+`false` into `terminable`, but only one will see `true` out of it. This acts as
+an indicator for which function may continue, where the other may have to take
+special care as to not leave the instance in a state that would be dangerous to
+signal.
+
+#### Guest acquires `terminable`
+
+The guest exits in the next handful of instructions. The `KillSwitch` that
+failed to acquire the `true` in `terminable` exits with an indication that it
+could not terminate the guest. In this circunstance, we are certain that there
+is a non-conditional and short path out of the guest comprising a return from
+`exit_guest_region` and a context swap back to the host code. Timeouts
+"failing" due to this are only failing because the guest is about to exit, and
+a signal would have no interesting additional benefit.
+
+#### `KillSwitch` acquires `terminable`
+
+In a more unfortunate circumstance, the `KillSwitch` is what observed the
+`true` out of `terminable`. In this case, the guest observed `false` and must
+not proceed, so that whenever an imminent `SIGALRM` from the corresponding
+`KillSwitch` arrives, it will be in a guaranteed-to-be-safe spin loop, or on
+its way there with only signal-safe state.
+
+The `KillSwitch` itself will signal the guest as any other `Domain::Guest`
+interruption.
+
+### `C -> B` timeout
+
+The case of a timeout while exiting from a hostcall is very similar to timeouts
+while entering a hostcall. Either the `KillSwitch` sees a guest in
+`Domain::Guest`, prohibits a state change, and signals the guest, or
+`KillSwitch` sees a guest in `Domain::Hostcall` and updates the guest to
+`Domain::Terminated`. In the latter case, the guest will be free to run when
+the `KillSwitch` returns, at which point it will have the same behavior as
+observing a timeout in any hostcall.
+
+### `C -> E` timeout, during hostcall terminating instance
+
+The `KillSwitch` that fires acquires `terminable` and then attempts to acquire
+a lock on `execution_domain`. The `KillSwitch` will see `Domain::Hostcall`, and
+will update to `Domain::Terminated`. The shared `KillState` will be not used by
+`lucet_runtime` again in the future, because after returning to the host it
+will be replaced by a new `KillState`.
+
+### `C -> E` timeout, during hostcall observing timeout
+
+A timeout while a hostcall is observing an earlier timeout will have no effect
+- a timeout must have occurred already, so the `KillSwitch` that fired will not
+acquire `terminable`, and will return without ceremony.
+
+### `D -> E` timeout
+
+A timeout while observing a cancelled guest will have no effect - a timeout
+must have occurred already, so the `KillSwitch` that fired will not acquire
+`terminable`, and will return without ceremony.
+
+### Timeout in hostcall fault
+
+As promised, a note about what happens when a timeout occurs directly when a
```

Comment:
Much sentence.  Please shorten.  This feels like the really important part about "sternly".  It seems that this is the scenario in which a KillSwitch cannot guarantee that an instance will be killed.  I'm not feeling confident that I understand this scenario.   I *think* it's this:

- A hostcall faults.  Domain::Hostcall.
- At the very same moment, a KillSwitch is invoked.
- In some cases, execution_domain gets updated to Domain:Terminated.

But then what happens?  I'm not sure what it means when you write, "no further effect will be had."  Does that mean the KillSwitch has no further effect, but the instance will live out the remainder of its lifecycle from `Domain::Terminated`?

Reply:
That's the right sequence of events for this section, yep! "no further effect will be had" refers to the fact that a hostcall fault happened, so one way or the other the instance will not resume, meaning the timeout will never be observed (which would require resuming and reaching `end_hostcall`)

I've expanded on that point and rephrased this for improved human-friendliness.